stable_baselines3/common/save_util.py

```diff
@@ -174,32 +176,61 @@ def json_to_data(json_string: str,
     return return_data
 
 
-def save_to_zip_file(save_path: str, data: Dict[str, Any] = None,
-                     params: Dict[str, Any] = None, tensors: Dict[str, Any] = None) -> None:
+@functools.singledispatch
+def open_path(path, verbose=0, suffix=None) -> io.IOBase:
+    # capture the most generic and assume it is correct
+    return path
+
+
+@open_path.register(str)
+def open_path_str(path: str, verbose=0, suffix=None) -> io.IOBase:
+    return open_path(pathlib.Path(path), verbose=verbose, suffix=suffix)
+
+
+@open_path.register(pathlib.Path)
+def open_path_pathlib(path: pathlib.Path, verbose=0, suffix=None) -> io.IOBase:
+    try:
+        if suffix and path.suffix != f".{suffix}":
+            path = pathlib.Path(f"{path}.{suffix}")
+        if path.exists() and not path.is_dir() and verbose == 2:
+            warnings.warn(f"Path {path} exists, will overwrite it.")
+        path = path.open("wb")
+    except IsADirectoryError:
+        warnings.warn(f"Path {path} is a folder. Will save instead to {path}_2")
+        path = pathlib.Path(f"{path}_2")
+    except FileNotFoundError:  # Occurs when the parent folder doesn't exist
+        warnings.warn(f"Path {path.parent} does not exist. Will create it.")
+        path.parent.mkdir(exist_ok=True, parents=True)
+
+    # if opening was successful uses the identity function
+    # if opening failed with IsADirectory|FileNotFound, calls open_path_pathlib
+    #   with corrections
+    return open_path(path)
+
+
+def save_to_zip_file(save_path, data: Dict[str, Any] = None,
+                     params: Dict[str, Any] = None, tensors: Dict[str, Any] = None, verbose=0) -> None:
     """
     Save a model to a zip archive.
 
-    :param save_path: Where to store the model.
+    :param save_path: (Union[str, pathlib.Path, io.IOBase]) Where to store the model.
```

Comment:
I think we could replace io.IOBase with io.BytesIO all around the code. I think all saving is done in bytes. This way we would not need casting :)